src/ir/analysis/derive_copy.rs

```diff
@@ -0,0 +1,416 @@
+    // only include the edges from the IR graph that can affect whether a type
+    // can derive copy or not.
+    dependencies: HashMap<ItemId, Vec<ItemId>>,
+}
+
+impl<'ctx, 'gen> CannotDeriveCopy<'ctx, 'gen> {
+    fn consider_edge(kind: EdgeKind) -> bool {
+        match kind {
+            // These are the only edges that can affect whether a type can derive
+            // copy or not.
+            EdgeKind::BaseMember |
+            EdgeKind::Field |
+            EdgeKind::TypeReference |
+            EdgeKind::VarType |
+            EdgeKind::TemplateArgument |
+            EdgeKind::TemplateDeclaration |
+            EdgeKind::TemplateParameterDefinition => true,
+
+            EdgeKind::Constructor |
+            EdgeKind::Destructor |
+            EdgeKind::FunctionReturn |
+            EdgeKind::FunctionParameter |
+            EdgeKind::InnerType |
+            EdgeKind::InnerVar |
+            EdgeKind::Method => false,
+            EdgeKind::Generic => false,
+        }
+    }
+
+    fn insert(&mut self, id: ItemId) -> ConstrainResult {
+        trace!("inserting {:?} into the cannot_derive_copy set", id);
+
+        let was_not_already_in_set = self.cannot_derive_copy.insert(id);
+        assert!(
+            was_not_already_in_set,
+            "We shouldn't try and insert {:?} twice because if it was \
+             already in the set, `constrain` should have exited early.",
+            id
+        );
+
+        ConstrainResult::Changed
+    }
+
+    fn insert_array(&mut self, id: ItemId) -> ConstrainResult {
+        trace!("inserting {:?} into the cannot_derive_copy_in_array set", id);
+
+        let was_not_already_in_set = self.cannot_derive_copy_in_array.insert(id);
+        assert!(
+            was_not_already_in_set,
+            "We shouldn't try and insert {:?} twice because if it was \
+             already in the set, `constrain` should have exited early.",
+            id
+        );
+
+        ConstrainResult::Changed
+    }
+}
+
+impl<'ctx, 'gen> MonotoneFramework for CannotDeriveCopy<'ctx, 'gen> {
+    type Node = ItemId;
+    type Extra = &'ctx BindgenContext<'gen>;
+    type Output = (HashSet<ItemId>, HashSet<ItemId>);
+
+    fn new(ctx: &'ctx BindgenContext<'gen>) -> CannotDeriveCopy<'ctx, 'gen> {
+        let cannot_derive_copy = HashSet::new();
+        let cannot_derive_copy_in_array = HashSet::new();
+        let dependencies = generate_dependencies(ctx, Self::consider_edge);
+
+        CannotDeriveCopy {
+            ctx,
+            cannot_derive_copy,
+            cannot_derive_copy_in_array,
+            dependencies,
+        }
+    }
+
+    fn initial_worklist(&self) -> Vec<ItemId> {
+        self.ctx.whitelisted_items().iter().cloned().collect()
+    }
+
+    fn constrain(&mut self, id: ItemId) -> ConstrainResult {
+        trace!("constrain: {:?}", id);
+
+        if self.cannot_derive_copy.contains(&id) || self.cannot_derive_copy_in_array.contains(&id) {
```

Comment:
Yes. I do want to make sure they insert into both sets. Otherwise, I don't have a way to make sure the function reaches a fix-point. I can't check that they are in both sets as there're elements only belongs to the second set. From what I see in the trace if it inserts in the first set, it also needs to insert into the second set. I split it for the situation that it doesn't insert in the first set and insert into the second set (things related to Named both explicitly and transitively).